### PR TITLE
Build update option

### PR DESCRIPTION
**Pull request by ikaneshiro**
This PR changes the `conveyor` interface to accept a `bundle` instead of a recipe to allow the build system to be run over a `bundle`. 

Builds occur as they have been before where if there is an existing container at the destination, there is an option to remove it or stop the build. The `--force` option defaults to remove the container at the destination.

This PR adds an `--update` option that skips the `conveyorPacker` portion, allowing builds with definitions to be run over an existing sandbox

### CODE REVIEW

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 1]
```diff
@@ -77,31 +59,50 @@ func (a *SandboxAssembler) Assemble(b *types.Bundle, path string) (err error) {
...
-	err := ioutil.WriteFile(filepath.Join(b.Rootfs(), "/.singularity.d/env/90-environment.sh"), []byte("#!/bin/sh\n\n"+b.Recipe.ImageData.Environment+"\n"), 0775)
-	return err
-}
+func insertDefinition(b *types.Bundle) error {
 
-func insertRunScript(b *types.Bundle) error {
-	err := ioutil.WriteFile(filepath.Join(b.Rootfs(), "/.singularity.d/runscript"), []byte("#!/bin/sh\n\n"+b.Recipe.ImageData.Runscript+"\n"), 0775)
-	return err
-}
+	//if update, check for existing definition and move it to bootstrap history
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `// if update, check for existing definition and move it to bootstrap history`

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 2]
```diff
@@ -77,31 +59,50 @@ func (a *SandboxAssembler) Assemble(b *types.Bundle, path string) (err error) {
...
+	if b.Update {
+		if _, err := os.Stat(filepath.Join(b.Rootfs(), "/.singularity.d/Singularity")); err == nil {
+			//make bootstrap_history directory if it doesnt exist
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `// make bootstrap_history directory if it doesnt exist`

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 3]
```diff
@@ -77,31 +59,50 @@ func (a *SandboxAssembler) Assemble(b *types.Bundle, path string) (err error) {
...
+				if err != nil {
+					return err
+				}
+			}
 
-func insertStartScript(b *types.Bundle) error {
-	err := ioutil.WriteFile(filepath.Join(b.Rootfs(), "/.singularity.d/startscript"), []byte("#!/bin/sh\n\n"+b.Recipe.ImageData.Startscript+"\n"), 0775)
-	return err
-}
+			//look at number of files in bootstrap_history to give correct file name
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 4]
```diff
@@ -77,31 +59,50 @@ func (a *SandboxAssembler) Assemble(b *types.Bundle, path string) (err error) {
...
+			files, err := ioutil.ReadDir(filepath.Join(b.Rootfs(), "/.singularity.d/bootstrap_history"))
 
-func insertTestScript(b *types.Bundle) error {
-	err := ioutil.WriteFile(filepath.Join(b.Rootfs(), "/.singularity.d/test"), []byte("#!/bin/sh\n\n"+b.Recipe.ImageData.Test+"\n"), 0775)
-	return err
-}
+			//name is "Singularity" concatinated with an index based on number of other files in bootstrap_history
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 5]
```diff
@@ -77,31 +59,50 @@ func (a *SandboxAssembler) Assemble(b *types.Bundle, path string) (err error) {
...
+			len := strconv.Itoa(len(files))
 
-func insertDefinition(b *types.Bundle) error {
+			histName := "Singularity" + len
+
+			//move old definition into bootstrap_history
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 6]
```diff
@@ -119,11 +120,58 @@ func insertDefinition(b *types.Bundle) error {
...
-	if err != nil {
-		return err
+	if b.RunSection("labels") && len(b.Recipe.ImageData.Labels) > 0 {
+		sylog.Infof("Adding labels")
+
+		var text []byte
+
+		if _, err := os.Stat(filepath.Join(b.Rootfs(), "/.singularity.d/labels.json")); err == nil {
+			existingLabels := make(map[string]string)
+			//check for labels that already exist
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 7]
```diff
@@ -119,11 +120,58 @@ func insertDefinition(b *types.Bundle) error {
...
+			if err != nil {
+				return err
+			}
+
+			err = json.Unmarshal(jsonBytes, &existingLabels)
+			if err != nil {
+				return err
+			}
+
+			//add new labels to new map and check for collisions
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 8]
```diff
@@ -119,11 +120,58 @@ func insertDefinition(b *types.Bundle) error {
...
+			for key, value := range b.Recipe.ImageData.Labels {
+				if _, ok := existingLabels[key]; ok {
+					//overwrite collision if force flag is set
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/assemblers/assembler_sandbox.go [hunk 9]
```diff
@@ -119,11 +120,58 @@ func insertDefinition(b *types.Bundle) error {
...
+					if b.Force {
+						existingLabels[key] = value
+					} else {
+						sylog.Warningf("Label: %s already exists and force option is false, not overwriting", key)
+					}
+				}
+			}
+
+			//make new map into json
```
Comment (ArangoGutierrez) [inline src/pkg/build/assemblers/assembler_sandbox.go]: `//` + space

File: src/pkg/build/build.go [hunk 10]
```diff
@@ -326,6 +371,26 @@ func makeDef(spec string) (types.Definition, error) {
...
 
+func (b *Build) addOptions() {
+	b.b.Update = b.update
+	b.b.Force = b.force
+	b.b.NoTest = b.noTest
+	b.b.Sections = b.sections
+}
+
+// RunSection determines if a section name was specified
+func (b Build) runPre() bool {
```
Comment (ArangoGutierrez) [inline src/pkg/build/build.go]: func name != comment name

File: src/pkg/build/sources/conveyorPacker_arch_test.go [hunk 11]
```diff
@@ -35,14 +35,20 @@ func TestArchConveyor(t *testing.T) {
 	}
 	defer defFile.Close()
 
-	def, err := types.ParseDefinitionFile(defFile)
+	//create bundle to build into
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_arch_test.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_arch_test.go [hunk 12]
```diff
@@ -63,14 +69,20 @@ func TestArchPacker(t *testing.T) {
...
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_arch_test.go]: `//` + space

File: src/pkg/sylog/sylog.go [hunk 13]
```diff
@@ -167,3 +169,13 @@ func GetLevel() int {
 func GetEnvVar() string {
 	return fmt.Sprintf("SINGULARITY_MESSAGELEVEL=%d", loggerLevel)
 }
+
+// Writer returns an io.Writer to pass to an external packages logging utility. For example,
+// when --quiet option is set, this function returns ioutil.Discard writer to ignore output
```
Comment (ArangoGutierrez) [inline src/pkg/sylog/sylog.go]:
```go
// Writer returns an io.Writer to pass to an external packages logging utility. 
// i.e when --quiet flag is set, this function returns ioutil.Discard writer to ignore output
```

File: src/runtime/engines/imgbuild/process.go [hunk 14]
```diff
@@ -18,37 +19,58 @@ import (
...
+			sylog.Fatalf("failed to start %%post proc: %v\n", err)
+		}
+		if err := post.Wait(); err != nil {
+			sylog.Fatalf("post proc: %v\n", err)
+		}
 	}
-	if err := post.Wait(); err != nil {
-		sylog.Fatalf("post proc: %v\n", err)
+
+	//append environment
```
Comment (ArangoGutierrez) [inline src/runtime/engines/imgbuild/process.go]: `//` + space

File: src/runtime/engines/imgbuild/process.go [hunk 15]
```diff
@@ -18,37 +19,58 @@ import (
...
-		test.Stderr = os.Stderr
-
-		sylog.Infof("Running %%test script\n")
-		if err := test.Start(); err != nil {
-			sylog.Fatalf("failed to start %%test proc: %v\n", err)
-		}
-		if err := test.Wait(); err != nil {
-			sylog.Fatalf("test proc: %v\n", err)
+
+	//insert startscript
```
Comment (ArangoGutierrez) [inline src/runtime/engines/imgbuild/process.go]: `// `+ space

File: src/runtime/engines/imgbuild/process.go [hunk 16]
```diff
@@ -18,37 +19,58 @@ import (
...
+	if err := e.insertStartScript(); err != nil {
+		return fmt.Errorf("While inserting startscript: %v", err)
+	}
+
+	//insert runscript
```
Comment (ArangoGutierrez) [inline src/runtime/engines/imgbuild/process.go]: `// `+ space

File: src/runtime/engines/imgbuild/process.go [hunk 17]
```diff
@@ -18,37 +19,58 @@ import (
...
+	if err := e.insertRunScript(); err != nil {
+		return fmt.Errorf("While inserting runscript: %v", err)
+	}
+
+	//insert test script
```
Comment (ArangoGutierrez) [inline src/runtime/engines/imgbuild/process.go]: `// `+ space

**Review by ArangoGutierrez — DISMISSED** [review]

File: src/cmd/singularity/cli/build.go [hunk 18]
```diff
@@ -90,41 +92,40 @@ var BuildCmd = &cobra.Command{
...
-func checkBuildTargetCollision(path string, force bool) bool {
-	if _, err := os.Stat(path); err == nil {
-		//exists
-		if force {
-			os.RemoveAll(path)
-		} else {
+func checkBuildTarget(path string) bool {
+	if f, err := os.Stat(path); err == nil {
+		if update && !f.IsDir() {
+			sylog.Errorf("Only sandbox updating is supported.")
```
Comment (bauerm97) [inline src/cmd/singularity/cli/build.go]: `sylog.Fatalf` otherwise we keep going

File: src/pkg/build/build.go [hunk 19]
```diff
@@ -50,32 +58,46 @@ type Build struct {
...
 		return nil, fmt.Errorf("unable to parse JSON: %v", err)
 	}
 
-	return newBuild(def, dest, format)
+	return newBuild(def, dest, format, force, update, sections, noTest)
 }
 
-func newBuild(d types.Definition, dest, format string) (*Build, error) {
+func newBuild(d types.Definition, dest, format string, force, update bool, sections []string, noTest bool) (*Build, error) {
+	var err error
```
Comment (bauerm97) [inline src/pkg/build/build.go]: In here should do:

```
if update {
    format = "sandbox"
}
```

so users don't have to run `singularity build --update --sandbox ...`. It's already checked for in `cli/build.go` that the destination is a directory, so it's already implied.

**Review by bauerm97 — CHANGES_REQUESTED** [review]

**Review by bauerm97 — APPROVED** [review]

**Pull request merged**